Zend/zend_compile.c

```diff
@@ -9623,11 +9623,12 @@ static void zend_compile_encaps_list(znode *result, zend_ast *ast) /* {{{ */
 	last_const_node.op_type = IS_UNUSED;
 	for (i = 0; i < list->children; i++) {
 		zend_ast *encaps_var = list->child[i];
+
 		if (encaps_var->attr & (ZEND_ENCAPS_VAR_DOLLAR_CURLY|ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR)) {
-			if (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR) {
-				zend_error(E_DEPRECATED, "Using ${expr} (variable variables) in strings is deprecated, use {${expr}} instead");
-			} else {
+			if (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY) {
 				zend_error(E_DEPRECATED, "Using ${var} in strings is deprecated, use {$var} instead");
+			} else if (encaps_var->kind == ZEND_AST_VAR) {
```

Comment:
Move this to the outer if? Both `ZEND_ENCAPS_VAR_DOLLAR_CURLY` and `ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR` should only be set on `ZEND_AST_VAR`. I think the only line that needs to change is 9627.

```c
if (encaps_var->kind == ZEND_AST_VAR && (encaps_var->attr & (ZEND_ENCAPS_VAR_DOLLAR_CURLY|ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR))) {
```

Reply:
No, I'm doing this deliberately - ZEND_ENCAPS_VAR_DOLLAR_CURLY can be set both on AST_DIM and AST_VAR ([var_dump](http://www.php.net/var_dump)("${foo['bar']}");)

Aside: `"Using ${var} in strings is deprecated, use {$var} instead"` could be changed to `var[expr]` in an error message for the kind AST_DIM but possibly out of scope

```
	|	T_DOLLAR_OPEN_CURLY_BRACES expr '}'
			{ $$ = zend_ast_create_ex(ZEND_AST_VAR, ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR, $2); }
	|	T_DOLLAR_OPEN_CURLY_BRACES T_STRING_VARNAME '}'
			{ $$ = zend_ast_create_ex(ZEND_AST_VAR, ZEND_ENCAPS_VAR_DOLLAR_CURLY, $2); }
	|	T_DOLLAR_OPEN_CURLY_BRACES T_STRING_VARNAME '[' expr ']' '}'
			{ $$ = zend_ast_create_ex(ZEND_AST_DIM, ZEND_ENCAPS_VAR_DOLLAR_CURLY,
			      zend_ast_create(ZEND_AST_VAR, $2), $4); }
```

Reply:
Ah, you're right of course. In that case it might be better to actually disambiguate the flags and maybe then `ZEND_ASSERT(encaps_var->kind == ZEND_AST_VAR);`. Sorry for the confusion.

Reply:
See https://github.com/nikic/php-ast/pull/227/files#diff-e3090d0e11ee031e6cd05b5fe11454bfacc69057e2d022bb9254b1e0f32e63cd (different repo using php-ast) for examples of those

There's also `"{$a->b()}"` - but AST_METHOD_CALL and AST_NULLSAFE_METHOD_CALL and AST_PROP and AST_NULLSAFE_PROP currently don't have flags set on parsing (more than I initially thought).

Though it'd be clearer to explicitly spell out the ast kinds being checked in case we ever set flags in the future

Reply:
Ok, so we should probably then check for `VAR` or `DIM` in the outer `if` in case `METHOD_CALL` and `PROP` ever add any flags?

Reply:
Done

Reply:
I still think it would be more logical to move the `(encaps_var->kind == ZEND_AST_VAR || encaps_var->kind == ZEND_AST_DIM)` to the first if, as otherwise `encaps_var->attr & (ZEND_ENCAPS_VAR_DOLLAR_CURLY|ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR)` might succeed by complete coincidence of other flags flashing with those. Other than that the PR looks fine to me. It might still be better to un-clash the `ZEND_DIM_*` and `ZEND_ENCAPS_VAR_*` flags. I know I said the opposite before but I forgot that the flag can also be set on the `ZEND_AST_DIM`.

Reply:
I added the check of the kind in both the inner `if` and `else if` the subsequent commit - https://github.com/php/php-src/pull/9264/commits/c4180f2e4d9f39abc851c478536c1c71d0ecb000

```c
			if ((encaps_var->kind == ZEND_AST_VAR || encaps_var->kind == ZEND_AST_DIM) && (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY)) {
				zend_error(E_DEPRECATED, "Using ${var} in strings is deprecated, use {$var} instead");
			} else if (encaps_var->kind == ZEND_AST_VAR && (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR)) {
				zend_error(E_DEPRECATED, "Using ${expr} (variable variables) in strings is deprecated, use {${expr}} instead");
```

Reply:
I was thinking:
```c
if ((encaps_var->kind == ZEND_AST_VAR || encaps_var->kind == ZEND_AST_DIM) && (encaps_var->attr & (ZEND_ENCAPS_VAR_DOLLAR_CURLY|ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR))) {
	if (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY) {
		zend_error(E_DEPRECATED, "Using ${var} in strings is deprecated, use {$var} instead");
	} else {
		ZEND_ASSERT(encaps_var->kind == ZEND_AST_VAR);
		zend_error(E_DEPRECATED, "Using ${expr} (variable variables) in strings is deprecated, use {${expr}} instead");
	}
}
```
Anyway, the current way looks correct too so.

Reply:
Your suggestion would reintroduce the original bug and the ZEND_ASSERT would fail in debug builds `} else if (encaps_var->kind == ZEND_AST_VAR && (encaps_var->attr & ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR)) {`

AST_DIM can have the flag `1<<`.

Anyway, my preference was a tradeoff for readability and to limit the number of places to look at to understand the `kind` and `attr` combinations - the outer `->attr &` continues to be the unlikely branch that would usually be false

> The ast node flag constants ZEND_DIM_ALTERNATIVE_SYNTAX and
ZEND_ENCAPS_VAR_DOLLAR_CURLY_VAR_VAR node previously had identical values (1<<1),

Reply:
I also updated the PR description to match the latest commits